crates/bls-gadgets/src/y_to_bit.rs

```diff
@@ -17,36 +20,46 @@ use std::{marker::PhantomData, ops::Neg};
 /// range). Then we check that the cast element is <= half, which enforces that
 /// originally they were > half. For points in G2, we also check the
 /// lexicographical ordering.
-pub struct YToBitGadget<P: Bls12Parameters> {
-    parameters_type: PhantomData<P>,
+pub trait YToBitGadget<P: Bls12Parameters> {
+    fn y_to_bit(&self) -> Result<Boolean<P::Fp>, SynthesisError>;
+}
+
+pub trait FpUtils<F: PrimeField> {
+    fn is_eq_zero(
+        &self,
+    ) -> Result<Boolean<F>, SynthesisError>; 
+    fn normalize(
+        &self,
+    ) -> Result<Boolean<F>, SynthesisError>;
 }
 
-impl<P: Bls12Parameters> YToBitGadget<P> {
-    pub fn y_to_bit_g1<CS: ConstraintSystem<P::Fp>>(
-        mut cs: CS,
-        pk: &G1Gadget<P>,
-    ) -> Result<Boolean, SynthesisError> {
-        let y_bit = Self::normalize(&mut cs.ns(|| "g1 normalize"), &pk.y)?;
+impl YToBitGadget<Bls12_377_Parameters> for G1Var<Bls12_377_Parameters> 
+{
+    fn y_to_bit(
+        &self,
+    ) -> Result<Boolean<<Bls12_377_Parameters as Bls12Parameters>::Fp>, SynthesisError> {
+        let y_bit = self.y.normalize()?;
         Ok(y_bit)
     }
+}
 
-    pub fn y_to_bit_g2<CS: ConstraintSystem<P::Fp>>(
-        mut cs: CS,
-        pk: &G2Gadget<P>,
-    ) -> Result<Boolean, SynthesisError> {
+impl YToBitGadget<Bls12_377_Parameters> for G2Var<Bls12_377_Parameters> {
```

Comment:
Same comment as above - do we have to be tied to BLS12-377?

Reply:
Same answer